packages/api-client-core/src/GadgetFunctions.ts

```diff
@@ -206,3 +207,14 @@ export interface GlobalActionFunction<VariablesT> {
 
 export type AnyActionFunction = ActionFunctionMetadata<any, any, any, any, any, any> | GlobalActionFunction<any>;
 export type AnyBulkActionFunction = ActionFunctionMetadata<any, any, any, any, any, true>;
+
+export interface ComputedViewFunction<VariablesT extends VariablesOptions, SelectionsT extends FieldSelection, ResultT> {
+  (variables: VariablesT): Promise<ResultT>;
```

Comment:
(variables: VariablesT, options?: { select?: SelectionsT }): Promise<Select<ResultT, DefaultSelection<DefaultSelectionsT, SelectionsT, DefaultSelectionsT>>>

Reply:
@airhorns so I gave this a go and it seems I have to declare this as follows for TS to stop complaining about things:
```ts
export interface ComputedViewFunction<VariablesT, DefaultSelectionsT extends FieldSelection, SelectionsT extends DefaultSelectionsT, ResultT> {
  (variables: VariablesT, options?: { select?: SelectionsT }): Promise<Select<ResultT, DefaultSelection<DefaultSelectionsT, { select?: SelectionsT | null }, DefaultSelectionsT>>>;
  /// ...
}
```
The `extends FieldSelection` bit is required because `Select<a,b>` requires b to extend FieldSelection. So far so good. The part that doesn't make sense to me is that I had to do `SelectionsT extends DefaultSelectionsT` because `DefaultSelection<a,b,c>` requires b and c to `extend a`. But that seems backwards, DefaultSelectionsT has the full selection details and SelectionT has some subset so it should be `DefaultSelectionT extends SelectionT` no? But if I do that I can't make DefaultSelection<> to work. This is all very meta and makes my head hurt, so quite likely I'm misunderstanding something.

Reply:
Hm, maybe I misunderstood what the DefaultSelectionT type should be, looking at this example https://github.com/gadget-inc/js-clients/blob/main/packages/api-client-core/spec/TestSchema.ts#L75. It seems that DefaultSelectionT should have all fields, but optional. So in that sense SelectionT could be extension of DefaultSelectionT. I thought DefaultSelectionT would have all fields required.

Reply:
Thats correct -- the existing default selections have everything optional but I think thats really just happenstance more than anything, no strong reason they are actually that way. The *actual* selection type has to have all the keys optional so that users can pass only some keys and not have to exhaustively list `false` for everything else and I bet the default just matches that cause that was our headspace at the . Because they are out in the wild I think any new ones you produce should match methinks 👍

Reply:
I think with my improved understanding and leaning heavily on the TestSchema example I'm inclined to settle on this:
```ts
export interface ComputedViewFunction<
  Variables,
  AvailableSelection extends FieldSelection,
  Result
> {
  (variables: Variables, options?: { select?: AvailableSelection }): Promise<
    Select<Result, DefaultSelection<AvailableSelection, Selectable<AvailableSelection>, Required<AvailableSelection>>>
  >;
    // ...
}
```
I was struggling with the two Selection types, given that we can't really know what the actual selection is going to be at runtime. So I think it's fine to say that if the select: option is present the value is of type AvailableSelection and the default is a Required<> derivative of AvailableSelection. Does that make sense?

Reply:
Hrm, still something off, this give me `never`
```
Select<TotalInStock, DefaultSelection<AvailableTotalInStockSelection, Selectable<AvailableTotalInStockSelection>, Required<AvailableTotalInStockSelection>>>
```

Reply:
I'm thinking the issue is with the default arg `Required<AvailableSelection>`, we don't want just required fields, we want them all set to true, so I need a type that has all fields and all typed as 'true' I think.